qa/rpc-tests/feature_blocksign.py

```diff
@@ -0,0 +1,124 @@
+    assert(num_nodes >= required_signers)
+    script = "{}".format(50 + required_signers)
+    for i in range(num_nodes):
+        k = keys[i]
+        script += "41"
+        script += k.get_pubkey().hex()
+    script += "{}".format(50 + num_nodes) # num keys
+    script += "ae" # OP_CHECKMULTISIG
+    print('signblockscript', script)
+    return script
+
+class BlockSignTest(test_framework.BitcoinTestFramework):
+
+    # Dynamically generate N keys to be used for block signing.
+    def init_keys(self, num_keys):
+        self.keys = []
+        self.wifs = []
+        for i in range(num_keys):
+            k = key.CECKey()
+            pk_bytes = hashlib.sha256(str(random.getrandbits(256)).encode('utf-8')).digest()
+            k.set_secretbytes(pk_bytes)
+            w = wif(pk_bytes)
+            print('k.get_privkey()', k.get_privkey())
+            print('k.get_pubkey()', k.get_pubkey())
+            print("generated key {}: \n  priv: {}\n  pub: {}\n  wif: {}".format(i+1,
+                k.get_privkey(),
+                k.get_pubkey(),
+                w))
+            self.keys.append(k)
+            self.wifs.append(wif(pk_bytes))
+
+    def __init__(self, num_nodes, required_signers):
+        super().__init__()
+        self.setup_clean_chain = True
+        self.num_nodes = num_nodes
+        self.init_keys(self.num_nodes)
+        self.required_signers = required_signers
+        signblockscript = make_signblockscript(num_nodes, required_signers, self.keys)
+        self.extra_args = [[
+            "-chain=blocksign",
+            # We can't validate pegins since we don't run the parent chain.
+            "-validatepegin=0",
+            "-signblockscript={}".format(signblockscript)
+        ]] * self.num_nodes
+
+    def setup_network(self, split=False):
+        self.nodes = util.start_nodes(self.num_nodes, self.options.tmpdir, self.extra_args)
+        # Have every node import its block signing private key.
+        for i in range(self.num_nodes):
+            self.nodes[i].importprivkey(self.wifs[i])
+            if i + 1 < self.num_nodes:
+                util.connect_nodes_bi(self.nodes, i, i + 1)
+            else:
+                util.connect_nodes_bi(self.nodes, 0, i)
+        self.is_network_split = False
+        self.sync_all()
+
+    def mine_block(self):
+        # mine block in round robin sense: depending on the block number, a node
+        # is selected to create the block, others sign it and the selected node
+        # broadcasts it
+        mineridx = self.nodes[0].getblockcount() % self.num_nodes # assuming in sync
+        miner = self.nodes[mineridx]
+
+        # miner makes a block
+        block = miner.getnewblockhex()
+
+        # collect required_signers signatures
+        sigs = []
+        for i in range(self.required_signers):
+            result = miner.combineblocksigs(block, sigs)
+            util.assert_equal(result["complete"], False)
```

Comment:
Please also do a submitblock, then check that it didn't move the chain forward. Belts and suspenders!